clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -209,7 +209,10 @@ public RecordAppendResult append(TopicPartition tp,
                     return appendResult;
                 }
 
-                MemoryRecordsBuilder recordsBuilder = recordsBuilder(buffer, maxUsableMagic);
+                // We have to pass `size` because the buffer returned from `free.allocate` could, in theory, be larger
+                // than the requested size if `deallocate(ByteBuffer buffer, int size)` is invoked with a buffer larger
+                // than the `size` parameter
```

Comment:
@junrao That's true, we check the size against `poolableSize` and `buffer.capacity`. Do you know why we have this method that takes 2 parameters? It seems pointless.